File: bot.py
from telegram import ReplyKeyboardMarkup, KeyboardButton, WebAppInfo
from telegram.ext import Updater, CommandHandler, MessageHandler, Filters
import json
import logging
import os
from urllib.parse import urlencode

# 设置日志
logging.basicConfig(
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s',
    level=logging.DEBUG
)
logger = logging.getLogger(__name__)

# 从环境变量获取 TOKEN，如果没有则使用默认值
TOKEN = os.getenv('TELEGRAM_BOT_TOKEN', '7582221284:AAGvtmNC5RmjSRcumethqzgWPkSTJRYHxQg')

# 验证 TOKEN 是否有效
if not TOKEN or len(TOKEN.split(':')) != 2:
    raise ValueError("Invalid token! Please check your TELEGRAM_BOT_TOKEN")

# 管理员 ID
ADMIN_ID = "7036647707"

def start(update, context):
    """处理 /start 命令"""
    # 获取用户信息
    user = update.effective_user
    user_info = (
        f"👤 用户信息:\n"
        f"ID: {user.id}\n"
        f"用户名: @{user.username if user.username else '无'}\n"
        f"姓名: {user.first_name} {user.last_name if user.last_name else ''}\n"
        f"语言: {user.language_code if user.language_code else '未知'}"
    )
    
    # 记录用户信息
    logger.info(f"New user started bot: {user_info}")
    
    # 发送用户信息给管理员
    if str(ADMIN_ID).isdigit():
        context.bot.send_message(
            chat_id=ADMIN_ID,
            text=f"新用户开始使用机器人:\n{user_info}"
        )
    
    # 创建自定义键盘
    keyboard = [
        ['1', '2', '3', '4', '5', '6', '7'],
        ['8', '9', '10', '11', '12', '13', '14'],
        ['15', '16', '17', '18', '19', '20'],
        ['🔥 购买广告', '➡️ 下一页'],
        ['🔥 IM体育: 1个有效即享55%-70%-可...'],
        ['🎭升元棋牌❤️ 贷盈利70%分成招商❤️...'],
        [KeyboardButton(text='🌐 打开投资平台', web_app=WebAppInfo(url=miniapp_url))],
        ['📋 官方简介', '📁 分类'],
        ['👤 我的', '💰 推广赚钱'],
        ['🔥 广告投放', '❓ 帮助']
    ]
    
    # 创建 ReplyKeyboardMarkup
    reply_markup = ReplyKeyboardMarkup(
        keyboard,
        resize_keyboard=True,
        one_time_keyboard=False,
        input_field_placeholder="请选择功能或输入消息"
    )
    
    # 发送欢迎消息
    update.message.reply_text(
        f"👋 欢迎使用投资平台机器人!\n"
        f"🎯 请点击下方按钮选择功能\n"
        f"💡 提示: 点击数字按钮查看详细信息\n\n"
        f"{user_info}",
        reply_markup=reply_markup
    )

def handle_message(update, context):
    """处理所有消息"""
    try:
        # 获取用户信息
        user = update.effective_user
        user_info = {
            'user_id': user.id,
            'username': user.username or '',
            'first_name': user.first_name or '',
            'last_name': user.last_name or '',
            'language': user.language_code or ''
        }
        print('1111111111111111111111111111111111111111111111')
        print(user_info)
        
        # 构建 miniapp URL，包含用户信息和按钮信息
        base_url = "https://wsk1576025821.github.io/telegram-investment-app"
        
        if update.message.text and update.message.text.isdigit():
            number = int(update.message.text)
            user_info['button_type'] = 'number'
            user_info['button_value'] = str(number)
        else:
            user_info['button_type'] = 'text'
            user_info['button_value'] = update.message.text or ''
        
        # 添加时间戳
        user_info['timestamp'] = update.message.date.strftime('%Y-%m-%d %H:%M:%S')

        # 创建带参数的 URL
        miniapp_url = f"{base_url}?{urlencode(user_info)}"

        # 创建自定义键盘，添加 miniapp 链接按钮
        keyboard = [
            ['1', '2', '3', '4', '5', '6', '7'],
            ['8', '9', '10', '11', '12', '13', '14'],
            ['15', '16', '17', '18', '19', '20'],
            ['🔥 购买广告', '➡️ 下一页'],
            ['🔥 IM体育: 1个有效即享55%-70%-可...'],
            ['🎭升元棋牌❤️ 贷盈利70%分成招商❤️...'],
            [KeyboardButton(text='🌐 打开投资平台', web_app=WebAppInfo(url=miniapp_url))],
            ['📋 官方简介', '📁 分类'],
            ['👤 我的', '💰 推广赚钱'],
            ['🔥 广告投放', '❓ 帮助']
        ]
        
        # 创建 ReplyKeyboardMarkup
        reply_markup = ReplyKeyboardMarkup(
            keyboard,
            resize_keyboard=True,
            one_time_keyboard=False,
            input_field_placeholder="请选择功能或输入消息"
        )
        
        # 记录所有收到的消息
        logger.info(f"Received message: {update.message.text}")
        
        # 检查是否是数字按钮点击
        if update.message.text and update.message.text.isdigit():
            number = int(update.message.text)
            if 1 <= number <= 20:
                message = (
                    f"🎯 您选择了选项 {number}\n"
                    f"⏳ 正在处理您的请求...\n"
                    f"📞 客服稍后会与您联系\n\n"
                    f"{user_info}"
                )
                update.message.reply_text(message)
                
                # 同时发送给管理员
                if str(ADMIN_ID).isdigit():
                    admin_message = (
                        f"🔔 用户点击了选项 {number}\n\n"
                        f"{user_info}\n"
                        f"📅 时间: {update.message.date.strftime('%Y-%m-%d %H:%M:%S')}"
                    )
                    context.bot.send_message(
                        chat_id=ADMIN_ID,
                        text=admin_message
                    )
                return

        # 处理其他按钮点击
        if update.message.text:
            message_with_info = None
            
            if "购买广告" in update.message.text:
                message_with_info = f"🔥 购买广告请联系客服\n\n{user_info}"
            elif "IM体育" in update.message.text:
                message_with_info = f"🏆 IM体育详情请联系客服\n\n{user_info}"
            elif "官方简介" in update.message.text:
                message_with_info = f"📋 官方简介\n\n这是一个专业的投资理财平台...\n\n{user_info}"
            elif "分类" in update.message.text:
                message_with_info = (
                    "📁 分类信息\n\n"
                    "1. 体育投资\n"
                    "2. 棋牌游戏\n"
                    "3. 电子竞技\n\n"
                    f"{user_info}"
                )
            elif "我的" in update.message.text:
                handle_profile_click(update, context)
            elif "推广赚钱" in update.message.text:
                message_with_info = f"💰 推广说明\n\n加入我们的推广计划，享受高额佣金\n\n{user_info}"
            elif "广告投放" in update.message.text:
                message_with_info = f"🔥 广告投放说明\n\n请联系客服了解详情\n\n{user_info}"
            elif "帮助" in update.message.text:
                message_with_info = f"❓ 帮助中心\n\n如有问题请联系在线客服\n\n{user_info}"
            
            if message_with_info:
                update.message.reply_text(message_with_info)
            
            # 移到这里，确保所有按钮点击都会发送管理员通知
            if str(ADMIN_ID).isdigit():
                admin_message = (
                    f"🔔 用户点击了: {update.message.text}\n\n"
                    f"{user_info}\n"
                    f"📅 时间: {update.message.date.strftime('%Y-%m-%d %H:%M:%S')}"
                )
                context.bot.send_message(
                    chat_id=ADMIN_ID,
                    text=admin_message
                )
        
    except Exception as e:
        logger.error(f"Error processing message: {e}", exc_info=True)

def handle_number_click(update, context, number):
    """处理数字按钮点击"""
    message = (
        f"🎯 您选择了选项 {number}\n"
        f"⏳ 正在处理您的请求...\n"
        f"📞 客服稍后会与您联系"
    )
    update.message.reply_text(message)

# 添加其他按钮处理函数
def handle_ad_click(update, context):
    update.message.reply_text("🔥 购买广告请联系客服")

def handle_sport_click(update, context):
    update.message.reply_text("🏆 IM体育详情请联系客服")

def handle_intro_click(update, context):
    update.message.reply_text("📋 官方简介\n\n这是一个专业的投资理财平台...")

def handle_category_click(update, context):
    update.message.reply_text("📁 分类信息\n\n1. 体育投资\n2. 棋牌游戏\n3. 电子竞技")

def handle_profile_click(update, context):
    """处理个人中心点击"""
    user = update.effective_user
    profile_info = (
        f"👤 个人信息\n\n"
        f"用户ID: {user.id}\n"
        f"用户名: @{user.username if user.username else '未设置'}\n"
        f"姓名: {user.first_name} {user.last_name if user.last_name else ''}\n"
        f"语言: {user.language_code if user.language_code else '未知'}\n\n"
        f"💫 账户状态: {'已认证' if str(user.id) == ADMIN_ID else '普通用户'}\n"
        f"📅 加入时间: {update.message.date.strftime('%Y-%m-%d %H:%M:%S')}\n\n"
        f"如需更多信息，请联系客服"
    )
    update.message.reply_text(profile_info)

def handle_promotion_click(update, context):
    update.message.reply_text("💰 推广说明\n\n加入我们的推广计划，享受高额佣金")

def handle_ad_post_click(update, context):
    update.message.reply_text("🔥 广告投放说明\n\n请联系客服了解详情")

def handle_help_click(update, context):
    update.message.reply_text("❓ 帮助中心\n\n如有问题请联系在线客服")

def error_handler(update, context):
    """处理错误"""
    logger.error(f"Update {update} caused error {context.error}")

def main():
    """启动机器人"""
    updater = Updater(TOKEN, use_context=True)
    dp = updater.dispatcher
    
    # 添加处理程序
    dp.add_handler(CommandHandler("start", start))
    # 使用通用消息处理器替代特定的 web_app_data 处理器
    dp.add_handler(MessageHandler(Filters.all, handle_message))
    
    # 错误处理
    dp.add_error_handler(error_handler)
    
    # 启动机器人
    updater.start_polling()
    logger.info("Bot started...")
    print("Bot is running! Press Ctrl+C to stop.")
    print("Send a message to the bot to get your Telegram ID")
    updater.idle()

if __name__ == '__main__':
    main()

File: bot.js
const TelegramBot = require('node-telegram-bot-api');
const dotenv = require('dotenv');
const fs = require('fs');
const lockFile = 'bot.lock';

// 加载环境变量
dotenv.config();

// Bot Token
const token = process.env.TELEGRAM_BOT_TOKEN || '7582221284:AAGvtmNC5RmjSRcumethqzgWPkSTJRYHxQg';
const adminId = "7036647707";

// 添加配置选项
const botOptions = {
    polling: {
        interval: 300,
        autoStart: true,
        params: {
            timeout: 10
        }
    },
    // 添加 webHook: false 确保不会同时使用 webhook
    webHook: false
};

// 创建 bot 实例时使用这些选项
const bot = new TelegramBot(token, botOptions);

// 基础 URL
const BASE_URL = encodeURI("https://nuxt-activity-dev.dx252.com//venue?id=zr&i-code=9079519");

// 创建键盘布局
const getKeyboard = (webAppUrl) => {
    console.log("Creating keyboard with WebApp URL:", webAppUrl);
    
    return {
        reply_markup: {
            keyboard: [
                ['1', '2', '3', '4', '5', '6', '7'],
                ['8', '9', '10', '11', '12', '13', '14'],
                ['15', '16', '17', '18', '19', '20'],
                ['📱 购买广告', '⬇️ 下一页'],
                ['🔥 IMA体育: 1个有效创手55%-76%-可...'],
                ['🔥手动挂单❤️ 资金利70%分成招商...'],
                [{
                    text: '🌐 打开投资平台',
                    web_app: {
                        url: webAppUrl
                    }
                }],
                ['📋 官方简介', '💳 分类'],
                ['👨‍💼 我的', '🔰 推广赚钱'],
                ['📱 广告投放', '❓ 帮助']
            ],
            resize_keyboard: true,
            one_time_keyboard: false
        }
    };
};

// 检查是否已经有实例在运行
function checkLock() {
    try {
        if (fs.existsSync(lockFile)) {
            const pid = fs.readFileSync(lockFile, 'utf8');
            try {
                // 检查进程是否还在运行
                process.kill(parseInt(pid), 0);
                console.error(`Bot is already running with PID ${pid}`);
                process.exit(1);
            } catch (e) {
                // 进程不存在，删除锁文件
                fs.unlinkSync(lockFile);
            }
        }
        // 创建新的锁文件
        fs.writeFileSync(lockFile, process.pid.toString());
    } catch (error) {
        console.error('Error handling lock file:', error);
    }
}

// 在启动时检查锁
checkLock();

// 在退出时删除锁文件
process.on('exit', () => {
    try {
        fs.unlinkSync(lockFile);
    } catch (error) {
        // 忽略错误
    }
});

// 创建 WebApp URL
const createWebAppUrl = (userInfo) => {
    const params = new URLSearchParams();
    
    // 确保所有参数都经过正确编码
    const safeUserInfo = {
        user_id: userInfo.user_id || '',
        username: encodeURIComponent(userInfo.username || 'anonymous'),
        first_name: encodeURIComponent(userInfo.first_name || ''),
        last_name: encodeURIComponent(userInfo.last_name || ''),
        language: userInfo.language || 'zh',
        chat_id: userInfo.chat_id || '',
        is_bot: String(userInfo.is_bot || false),
        is_premium: String(userInfo.is_premium || false),
        timestamp: encodeURIComponent(userInfo.timestamp || new Date().toISOString())
    };
    
    Object.entries(safeUserInfo).forEach(([key, value]) => {
        params.append(key, value);
    });
    
    return `${BASE_URL}?${params.toString()}`;
};

// 添加发送带图片和按钮的消息函数
async function sendPromotionalMessage(chatId) {
    try {
        // 构建用户参数
        const params = new URLSearchParams({
            source: 'promo',
            timestamp: new Date().toISOString()
        });

        const webAppUrl = `${BASE_URL}?${params.toString()}`;
        
        // 创建内联键盘按钮
        const inlineKeyboard = {
            reply_markup: {
                inline_keyboard: [
                    [{
                        text: '🌟 立即投资',
                        web_app: {
                            url: webAppUrl
                        }
                    }],
                    [{
                        text: '📱 分享给朋友',
                        url: `https://t.me/share/url?url=${encodeURIComponent(webAppUrl)}`
                    }]
                ]
            },
            parse_mode: 'HTML'
        };

        // 发送图片消息
        await bot.sendPhoto(chatId, 
            'path/to/your/promo/image.jpg', // 替换为您的图片路径或URL
            {
                caption: `
🔥 <b>投资平台最新优惠</b>

💰 高收益投资项目
✨ 专业团队管理
🔒 资金安全保障
                
<i>点击下方按钮立即开始！</i>`,
                parse_mode: 'HTML',
                ...inlineKeyboard
            }
        );

    } catch (error) {
        console.error('Error sending promotional message:', error);
    }
}

// 添加发送视频消息的函数
async function sendVideoMessage(chatId) {
    try {
        // 构建带参数的 URL
        const params = new URLSearchParams({
            source: 'video_promo',
            timestamp: new Date().toISOString()
        });
        const webAppUrl = `${BASE_URL}?${params.toString()}`;

        // 创建内联键盘按钮
        const messageOptions = {
            caption: `
🎉 恭喜老板爆奖了
🌟 祝您一路长虹

🔥🔥 恭喜玩家 展名大佬在 德信飞投 真得 1001000 USDT！🔥🔥
🎯💕 欢迎再进，一路长红❤️，再接再厉 🎯`,
            parse_mode: 'HTML',
            reply_markup: {
                inline_keyboard: [
                    [
                        {
                            text: '💎 德信飞投 💎',
                            web_app: {
                                url: webAppUrl
                            }
                        }
                    ],
                    [
                        {
                            text: '🎮 3D捕鱼',
                            callback_data: 'game_fish'
                        },
                        {
                            text: '🎲 电子游戏',
                            callback_data: 'game_slot'
                        }
                    ],
                    [
                        {
                            text: '🎯 香港六合彩',
                            callback_data: 'lottery_hk'
                        },
                        {
                            text: '🍁 加拿大28',
                            callback_data: 'lottery_ca'
                        }
                    ]
                ]
            }
        };

        // 发送视频消息
        await bot.sendVideo(
            chatId,
            'path/to/your/promo.mp4', // 替换为您的视频文件路径
            messageOptions
        );

        console.log('视频消息发送成功，时间:', new Date().toLocaleString());
    } catch (error) {
        console.error('发送视频消息失败:', error);
    }
}

// 修改定时发送函数
function startAutoMessage() {
    const chatId = adminId; // 使用 adminId 而不是环境变量
    
    if (!chatId) {
        console.error('未设置 adminId');
        return;
    }

    console.log('开始自动发送消息服务，目标chatId:', chatId);

    // 修改为每30秒发送一次
    setInterval(async () => {
        try {
            // 构建消息URL
            const params = new URLSearchParams({
                source: 'auto',
                timestamp: encodeURIComponent(new Date().toISOString())
            });

            const webAppUrl = `${BASE_URL}?${params.toString()}`;
            
            // 创建内联键盘按钮
            const messageOptions = {
                parse_mode: 'HTML',
                reply_markup: {
                    inline_keyboard: [
                        [{
                            text: '💎 德信飞投 💎',
                            web_app: {
                                url: webAppUrl
                            }
                        }],
                        [{
                            text: '🎮 3D捕鱼',
                            callback_data: 'game_fish'
                        },
                        {
                            text: '🎲 电子游戏',
                            callback_data: 'game_slot'
                        }],
                        [{
                            text: '🎯 香港六合彩',
                            callback_data: 'lottery_hk'
                        },
                        {
                            text: '🍁 加拿大28',
                            callback_data: 'lottery_ca'
                        }]
                    ]
                }
            };

            // 发送消息
            await bot.sendMessage(
                chatId,
                `
🎉 恭喜老板爆奖了
🌟 祝您一路长虹

🔥🔥 恭喜玩家 展名大佬在 德信飞投 真得 1001000 USDT！🔥🔥
🎯💕 欢迎再进，一路长红❤️，再接再厉 🎯`,
                messageOptions
            );

            console.log('自动消息发送成功，时间:', new Date().toLocaleString());

        } catch (error) {
            console.error('发送自动消息失败:', error);
        }
    }, 30000); // 30000ms = 30秒
}

// 在 bot 启动时开始自动发送
bot.on('polling_error', (error) => {
    console.error('Polling error:', error);
});

// 确保 bot 成功启动后再开始发送消息
bot.getMe().then((botInfo) => {
    console.log('Bot 启动成功:', botInfo.username);
    startAutoMessage();  // 直接启动自动发送消息
    console.log('自动发送消息服务已启动');
}).catch((error) => {
    console.error('Bot 启动失败:', error);
});

// 处理其他消息
bot.on('message', async (msg) => {
    try {
        // 检查是否有 WebApp 数据
        if (msg.web_app_data) {
            console.log('Received WebApp data:', msg.web_app_data);
            return;
        }

        const chatId = msg.chat.id;
        const user = msg.from;
        const text = msg.text;

        // 构建用户信息（与 /start 命令保持一致）
        const userInfo = {
            user_id: user.id.toString(), // 确保是字符串
            username: user.username || 'anonymous',
            first_name: user.first_name || '',
            last_name: user.last_name || '',
            language: user.language_code || 'zh',
            chat_id: chatId.toString(), // 确保是字符串
            is_bot: (user.is_bot || false).toString(), // 确保是字符串
            is_premium: (user.is_premium || false).toString(), // 确保是字符串
            timestamp: new Date().toISOString()
        };

        // 创建 WebApp URL
        const params = new URLSearchParams(userInfo);
        const webAppUrl = `${BASE_URL}?${params.toString()}`;
        console.log('Message handler - User Info:', userInfo);
        console.log('Message handler - Generated URL:', webAppUrl);

        // 使用 getKeyboard 函数创建键盘
        const keyboard = getKeyboard(webAppUrl);
        if (!keyboard) {
            console.error('Failed to create keyboard in message handler');
            return;
        }

        // 每次消息都更新键盘
        await bot.sendMessage(chatId, '请选择操作:', keyboard);

    } catch (error) {
        console.error('Error handling message:', error);
    }
});

// 添加错误处理
bot.on('polling_error', (error) => {
    // 如果是冲突错误，尝试重新启动
    if (error.code === 'ETELEGRAM' && error.response.statusCode === 409) {
        console.log('Conflict detected, restarting polling...');
        bot.stopPolling().then(() => {
            setTimeout(() => {
                bot.startPolling();
            }, 1000);
        });
    } else {
        console.error('Polling error:', error);
    }
});

// 添加优雅退出处理
process.on('SIGINT', () => {
    console.log('Stopping bot...');
    bot.stopPolling().then(() => {
        console.log('Bot stopped');
        process.exit(0);
    });
});

// 在 bot 初始化后添加这个函数
async function setupBot() {
    try {
        // 只设置命令列表
        await bot.setMyCommands([
            {
                command: 'start',
                description: '启动机器人'
            }
        ]);
        
        console.log('Bot commands setup completed');
    } catch (error) {
        console.error('Error setting up bot:', error);
    }
}

// 处理按钮回调
bot.on('callback_query', async (callbackQuery) => {
    const chatId = callbackQuery.message.chat.id;
    const data = callbackQuery.data;

    try {
        // 根据不同按钮处理不同逻辑
        switch (data) {
            case 'game_fish':
                await sendGameInfo(chatId, '3D捕鱼');
                break;
            case 'game_slot':
                await sendGameInfo(chatId, '电子游戏');
                break;
            case 'lottery_hk':
                await sendGameInfo(chatId, '香港六合彩');
                break;
            case 'lottery_ca':
                await sendGameInfo(chatId, '加拿大28');
                break;
        }
    } catch (error) {
        console.error('处理按钮回调失败:', error);
    }
});

// 发送游戏信息
async function sendGameInfo(chatId, gameName) {
    const params = new URLSearchParams({
        game: gameName,
        source: 'button_click',
        timestamp: new Date().toISOString()
    });

    const webAppUrl = `${BASE_URL}?${params.toString()}`;

    const messageOptions = {
        parse_mode: 'HTML',
        reply_markup: {
            inline_keyboard: [
                [{
                    text: `🎮 进入${gameName}`,
                    web_app: {
                        url: webAppUrl
                    }
                }]
            ]
        }
    };

    await bot.sendMessage(
        chatId,
        `
🎯 <b>${gameName}</b>

💫 热门游戏推荐
🎁 新人福利优惠
💰 上分无需等待

<i>点击下方按钮立即开始！</i>`,
        messageOptions
    );
}
